core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java

```diff
@@ -96,7 +99,8 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             if (initialRecoveryFilters != null  &&
                 RecoverySource.isInitialRecovery(shardRouting.recoverySource().getType()) &&
                 initialRecoveryFilters.match(node.node()) == false) {
-                return allocation.decision(Decision.NO, NAME, "node does not match index initial recovery filters [%s]",
+                return allocation.decision(Decision.NO, NAME,
+                    "initial allocation of the shrunken index is only allowed on nodes [%s] that hold a copy of every shard in the index",
```

Comment:
we could make the message dependent on the `shardRouting.recoverySource().getType()`. Use  the message you gave for index shrinking and a more general one o.w.?